// cng-node-js-utils/index.js
module.exports = {
  expHandlers: require("./express-handlers"),
  validators: require("./validators"),
  arrObj: require("./array-object"),
  dateUtil: require("./date-util"),
  excelUtils: require("./excel-utils"),
  expUtil: require("./express-util"),
  jsEncrypt: require("./jsencrypt"),
  LDAPLogin: require("./ldap-login"),
  myDevice: require("./my-device"),
  pdfUtil: require("./pdf-util"),
  postHandler: require("./post-handler"),
  qrCodeUtil: require("./qrcode-util"),
  rsaService: require("./rsa-service"),
  webService: require("./web-service"),
  Router: require("./express-handlers/router"),
  ExpressServer: require("./express-handlers/express-server"),
  secretUtil: require("./secret-util"),
  md5: require("./md5"),
  sendGmail: require("./send-gmail"),
  uniqueid: require("./uniqueid"),
  vnHandler: require("./vietnamese-handler"),
  waiting: require("./waiting"),
  fileFolderUtils: require("./file-folder-utils"),
  createSubFolder: require("./file-folder-utils").createSubFolder,
  convertPath2ObjectName: require("./file-folder-utils").convertPath2ObjectName,
  createEntryIndex: require("./file-folder-utils").createEntryIndex,
  captchaUtil: require("./captcha-util"),
  convertIpRange: require("./convert-ip-range"),
  // hàm delay tuần tự
  sleep: (milliseconds = 100) => {
    const date = Date.now();
    let currentDate = null;
    do {
      currentDate = Date.now();
    } while (currentDate - date < milliseconds);
  }
};
