// cng-node-js-utils/send-gmail.js
/**
 * Gửi email thông qua gmail,
 * Vào cấu hình trong gmail của google, cho phép xác thực và gửi mail từ hàm api, thì user sẽ có quyền gửi mail này
 * @param {*} receiver 
 * @param {*} subject 
 * @param {*} content 
 * @param {*} senderGmail 
 * @param {*} passGmail 
 */
module.exports = (receiver, subject, content, senderGmail, passGmail, fromEmail, host = 'smtp.gmail.com', port = 465, secure = true) => {
    return new Promise((resolve, reject) => {
        // cấu hình mail server kiểu của gmail default
        let configServer = {
            host,
            port,
            secure, // use SSL
            auth: {
                user: senderGmail,
                pass: passGmail
            }
        };

        // 
        let mailOptions = {
            from: fromEmail || configServer.auth.user || 'no_reply@gmail.com',
            to: receiver,
            subject: subject,
            html: content
            // text: content
        };

        let transporter = nodemailer.createTransport(configServer);

        transporter.sendMail(mailOptions, (error, info) => {
            if (error) {
                // console.log('error', error.message);
                reject(error.message)
            } else {
                // console.log('success', info);
                resolve(info.response);
            }
        });
    });
}
