// cng-node-js-utils/web-service.js
const axios = require('axios');
const url = require('url');
var FormData = require('form-data');

module.exports = {
    /**
     * Ham nay truy van mot lien ket de lay du lieu ve dang get
     * @param {*} urlLink Lien ket de lay du lieu
     * @param {*} token Token cung cap de xac thuc du lieu
     * @param {*} jsonData du lieu truy van dang json
     * @returns 
     * Ket qua tra ve la mot promise chua jsonData, headers va body
     */
    GET: (urlLink, token, jsonData) => {
        return axios({
            method: 'get',
            // chuyen doi json sang chuoi truy van neu co yeu cau du lieu truy van
            url: urlLink + (jsonData ? `?${new url.URLSearchParams(jsonData)}` : ``),
            headers: {
                'User-Agent': 'Cuongdq web service',
                'Authorization': token ? `Bearer ${token}` : undefined
            }
        })
            .then(res => ({
                headers: res.headers,
                body: res.data
            }))
            .catch(err => {
                console.log('ERROR-GET: ', err);
                return null;
            })
    },
    /**
     * Ham nay post du lieu kieu form binh thuong nhu submit cua web
     * @param {*} urlLink Lien ket de lay du lieu
     * @param {*} jsonData du lieu truy van dang json
     * @param {*} token Token cung cap de xac thuc du lieu
     * @param {*} headers Soan headers rieng de truy cap
     * @returns 
     */
    POST_HTTP: (urlLink, jsonData, token, headers) => {
        return axios.post(urlLink,
            JSON.stringify(jsonData),
            {
                headers: headers || {
                    'User-Agent': 'Cuongdq web service',
                    'Content-Type': 'application/x-www-form-urlencoded',
                    'Authorization': token ? `Bearer ${token}` : undefined,
                }
            })
            .then(res => ({
                headers: res.headers,
                body: res.data
            }))
            ;
    },
    /**
     * Ham nay post mot du lieu json len may chu
     * Kieu post mot chuoi json
     * @param {*} urlLink Lien ket de lay du lieu
     * @param {*} jsonData du lieu truy van dang json
     * @param {*} token Token cung cap de xac thuc du lieu
     * @param {*} headers Soan headers rieng de truy cap
     * @returns 
     */
    POST_JSON: (urlLink, jsonData, token, headers) => {
        return axios.post(urlLink,
            jsonData,
            {
                headers: headers || {
                    'User-Agent': 'Cuongdq web service',
                    'Content-Type': 'application/json',
                    'Authorization': token ? `Bearer ${token}` : undefined,
                }
            })
            .then(res => ({
                headers: res.headers,
                body: res.data
            }))
            ;
    },

    /**
     * Ham nay gui du lieu len server bang form_data
     * Kieu upload file len may chu file
     * @param {*} urlLink Lien ket de lay du lieu
     * @param {*} jsonData du lieu truy van dang json
     * @param {*} token Token cung cap de xac thuc du lieu
     * @returns 
     */
    POST_FORM: (urlLink, jsonData, token,) => {
        let form_data = new FormData();
        for (key in jsonData) {
            form_data.append(key, jsonData[key]);
        }
        return axios
            .post(urlLink,
                form_data,
                {
                    headers: {
                        'Content-Type': 'multipart/form-data',
                        ...form_data.getHeaders(),
                        'Authorization': token ? `Bearer ${token}` : undefined,
                    }
                })
            .then(res => ({
                headers: res.headers,
                body: res.data
            }))
            .catch(err => {
                console.log('ERROR-FORMDATA', err)
            });

    },
}
